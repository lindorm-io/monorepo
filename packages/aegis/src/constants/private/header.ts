import {
  EC_ENC_ALGORITHMS,
  EC_SIG_ALGORITHMS,
  OCT_ENC_DIR_ALGORITHMS,
  OCT_ENC_STD_ALGORITHMS,
  OCT_SIG_ALGORITHMS,
  OKP_ENC_ALGORITHMS,
  OKP_SIG_ALGORITHMS,
  RSA_ENC_ALGORITHMS,
  RSA_SIG_ALGORITHMS,
} from "@lindorm/kryptos";

export const TOKEN_HEADER_ALGORITHMS = [
  ...EC_ENC_ALGORITHMS,
  ...EC_SIG_ALGORITHMS,
  ...OKP_ENC_ALGORITHMS,
  ...OKP_SIG_ALGORITHMS,
  ...RSA_ENC_ALGORITHMS,
  ...RSA_SIG_ALGORITHMS,
  ...OCT_ENC_DIR_ALGORITHMS,
  ...OCT_ENC_STD_ALGORITHMS,
  ...OCT_SIG_ALGORITHMS,
] as const;

export const TOKEN_HEADER_TYPES = [
  // JOSE
  "JWE",
  "JWS",
  "JWT",

  // IANA
  "application/cwt",
  "application/cose; cose-type=cose-encrypt",
  "application/cose; cose-type=cose-mac",
  "application/cose; cose-type=cose-sign",
] as const;
