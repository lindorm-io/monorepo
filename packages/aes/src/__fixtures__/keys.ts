import { Kryptos, KryptosFromB64 } from "@lindorm/kryptos";

const defaults = {
  notBefore: new Date("2023-01-01T01:00:00.000Z"),
  updatedAt: new Date("2024-01-01T01:00:00.000Z"),
  expiresAt: new Date("2024-03-01T00:00:00.000Z"),
  issuer: "https://test.lindorm.io/",
  jwksUri: "https://test.lindorm.io/.well-known/jwks.json",
};

const EC: KryptosFromB64 = {
  ...defaults,
  id: "9e568fdb-ed87-5a24-800a-44e1a143d5da",
  curve: "P-521",
  createdAt: new Date("2023-01-01T00:00:00.000Z"),
  privateKey:
    "MIHuAgEAMBAGByqGSM49AgEGBSuBBAAjBIHWMIHTAgEBBEIATAYo3DQYroDV5EgJSUs_kOIvnEScZen73gXa5oQkub0ekQmgOJdPQjINUsdYRn67QK_oBdNUhVbtG_qdxqIgarehgYkDgYYABACN2PIVpRTdfXLmNxkg8Bk2m5netqYsNW2Lefhklr2jfJiVUJUDPoZoGfabGzHEgsKjP2HRbPsEI_tND3x4N9VW0QAID2UYXa7GN0izHWIFRdjVYuR5-0jywFtd-o_N2POdrvlV8xumdVK-TiSPEIdfKoL_Iu0e7IKTsJsj-UmE8rDJnw",
  publicKey:
    "MIGbMBAGByqGSM49AgEGBSuBBAAjA4GGAAQAjdjyFaUU3X1y5jcZIPAZNpuZ3ramLDVti3n4ZJa9o3yYlVCVAz6GaBn2mxsxxILCoz9h0Wz7BCP7TQ98eDfVVtEACA9lGF2uxjdIsx1iBUXY1WLkeftI8sBbXfqPzdjzna75VfMbpnVSvk4kjxCHXyqC_yLtHuyCk7CbI_lJhPKwyZ8",
  type: "EC",
};

export const TEST_EC_KEY = Kryptos.from("b64", {
  ...EC,
  algorithm: "ECDH-ES",
  operations: ["encrypt", "decrypt"],
  use: "enc",
});

const OCT: KryptosFromB64 = {
  ...defaults,
  id: "c124f42f-5a61-582d-9fdd-b834de3336a2",
  createdAt: new Date("2023-01-01T00:10:00.000Z"),
  privateKey: "NjVhNF9EK0BTUWJAd0AwUU9IQWkmdTgwWEo2QDJmYjI",
  type: "oct",
};

export const TEST_OCT_KEY = Kryptos.from("b64", {
  ...OCT,
  algorithm: "dir",
  operations: ["encrypt", "decrypt", "deriveKey"],
  use: "enc",
});

const RSA: KryptosFromB64 = {
  ...defaults,
  id: "8bb9bd86-2e70-5cd0-a1cc-b60a5eb6d574",
  createdAt: new Date("2023-01-01T00:30:00.000Z"),
  privateKey:
    "MIIJKQIBAAKCAgEApq21wR1t9ZnC0R2vgmUrDGjq0G5h5THrtC4kdKfavu2t7iwV0RHsQCJU_uboan63dQysFlYLrqvUluMKBgb8lNAGRc1M7LvGKh3zQlV1gtkh71oVlRqUpuGvY6FbBSCVZThC3oRXSlKpRXSMuHY_ohmsXSkZCDReTZevLmzTG7r_bDp8mK05moNU2dfBhh7KtTssyp6etufDrpxsBXF2i-2VzhwMX2GZJY_7DouXgHWTsLIOLqH18gNMUU8RYULNqiA8vBsubVWV9ygr6wLxdZslm4yAXfMVPj6dQ7FCwo4L7gxtqK3M5Olv1C7rRi0NpTCFuUZlOnufWcrSToPKRDvAFUkZPyk9qzyiww5ckVW3ZRaPpVF36RJN-VEU2dskfw2Md2NrMV3W8LNhXryU004CSjJy6bqmer2cQ9UwbZR8y9n4Vxf45lntZMaZ3ZOYnaXuHZL-j3AcZ82bRsApcQneRhJHElA-nJicuZ4s22-B_rvOBu8YNOR5PutAGZz7J5hGr8L-wGDHktqG0ChY30RRlbvK94XitwzXljhBeDmf5SRRoUVbba8PcF91c7IHlb5q4wWxOgEXM9kb0_Oik6iXCGfZVog8DB8uYI6P2Ta5pvMY_Jt_5U-ULu3P8MPNsfHtY0n1Kpf6MP6srdz2kZupT9VL_mReSv3korv7XHMCAwEAAQKCAgBSyOuBt6LCPfUHlSVzb9kiSEf8bx67TgCiVNqxgB9C_l0JcycA2NSNBCBQHJLqNfbbsYgVbfZOcrTbnnd_h9_S78F10OK4oAjsKUUFuNltPHhA3-7hSDr16hkJNH63PLhG9lssFvIeDZFaWGl-QukXE4ZLHRr8Orpt8n1FYzK1TrsHX8M7t6npJvqxWhqaCVgoDHSSTJdLBDDqXHHAgFrfZAkoPkEc7KY6hFkvH8EHQh6xP7aQjcvQwLeamNlqg9oPegSAb0WvczdRc0fg6WzRO1ZB_FZTk2UFLedBVFzysceWG-EpSu198r2i1wzvJXhmCAQbAi5MDlOWrwimi4mdbejnH_YbYJPvAnld5kpUgAvF3p-plAHaMHDA1hp8GrPcPh9QH5gmy5XEbdN7aV6ekAhMRIczxfWnzOc21l3nemK2bpGlSequBzQsAteYk5jYVBYOnfcTwzhS5nS1LPGJ_AVJEdyvJeM3C1WfC-nNuIdBPEvaQzeD8Re9-IY9Az_x_jW9HzeOzXfIkuNfqSyQeOmgUBIQRvyRg1Xfh0KQGp4Ihf2IcqajOo8-D8ZEUvHaheyEHdRXrORH0BFLIQ8UqmC4XjVOx7q3FURgXuqtFoGDx8cGvfHTLKozWrRhtS9DKCeEk-oGKBrv8cYF0-oebbSm2538Dji2XeobJWyldQKCAQEA13wEi30WTdjeMXVWUfgJkFfcApRMnXQHmtNDzS30qao96YNC5yPzrZyfFFhUoOWCi5fmOAMdMtS7BL6ZJp9aroMAi8IhccYdt8rEHxdbMR5pbQ_TqsH1wYW64ETFViR9rPsg5b8CvjmPxW-1FMwl5uOwmrC7sxUrCSTk1PglZVj1ALR2Z1rUjBKQLQFsh1iFBZepxGGD3kbeVqt04d_lYPSewyLGACvSx1BLbR5y8KBNVSialbb9n2sFKSL6gJytr-dHAMdPIM9SQGawrym0ZiEK5FxYsQUaVP5ulw0FrjzwafXBkFjl2FUjwFE3bnRhaR8BGbJAD8xqZarYSzhEHQKCAQEAxgSAqhB71wfDx57gxT8dnTzKNNsklBrMA3Cl3Zip2EG4L1pS1Qd_mL9LjIfF-zgWTMbGW_SQhoi1Fq4MSUFw5R30HP1qbcHw__WnzqLHYVRpJvIQGfe3LRx-EijblfAmKRWUsbvkn-bpwHrrJSm41I5jjmvam366t6NWLBiar4YAMZ-HTSv5-rJkJMsqGCRZHdregCMGSauIXddDZv-gr9OZnpQRX8CTe72tNdkOjGrknB6gxWJxcV3wU08uwUVwfiXDbqmT8SfFEtpAI_SKxXqDWCjJpKT-cxC5QesHapNEHQWO7M0MfsytejU6QmNTESRlb6P8Y5vlcTWMG8kdzwKCAQEAr2Tizici4HqFYPyBJnceUb8x9WeAdSjch8kuhvPBEdjgTBHrgvhggA_b7FN5jv1BJR6e5Tijq5w8HcCGKEnNJf67kwaQ8mqwBcmlmSkJsRFv_rXDNFpsDvkrAyD2_wb0xLgkLt0q_msb-zjRy_vfXfA277KrFTlFrku5dyAEUFn4EnrRZY8pMlppwDhIfmIu5D5ukMpO674De8hTewcjTafPEkZbpAgHqPRScJ84L2RLjeVe7j3kcr5JlRJDlCvztBosURXp8mrOKF1tsyXfEMIUe4wvJOdiZ4wxuAGBfjFb_uNqinPRoiSq1P9Gf0F7iHUUEDZmrnVUpSxKoe9KgQKCAQA4kJa0feryXPx3LJJmpgXuoavdd3A2goUewnIPIAn8haF4nWOmCPNRDewl468ReU8ddXa6U91pIOoh5WwLDQ-T03-lPFrHqawPYfwiketEiVlUT7d7w4Bm-Yml5Ff9t9Ee0VRdhR_k8ZB6dGJxhj6PMXGNAJA89PGffaAjDpA5bZzmj_yVded9zrG-tO58YP24Aoqmrdqg9sgIjnJpnMIpF9f9Lazc3mUC5wbwp-V5BRvy4OjqGZnwWXZKRgdEHHLrBhySowxoygSIwyGpY7S0fq_AYj5fsuYgtnlFND81GYZOiATw7g4QtqUjkMropfjXn77-Kijxx1m-gh5Y3ZWPAoIBAQCOmEBewD34QEnOioz0fZCzWnzd718k-ac7jbjbmTDM6dD6ZGS0RiG-6VfGqPy2D5wJvd3_pV68urq9pCcA6YErcXsXnRuGu1dX7Qu0-D-eHL7XsXYovzd3lbhfp31KP2wnbhwMnp81FO1wrHY9FvRHmIvWd7OE4z8JWDlspLJVqIFgcqOZJCHBYbz32nY8qkr2KhneJPJ4XibBje5xvRTiF_qTPAa633peDgoVCHkg8Ub1b6X9vg_VXSDeJy3EXlRYEGlfuhEfkSSg9oAYuI-XqnCGqj9JffrncWMq7TY5jduveOAyAZwk0FFsHrius3JUoV-aXlueJYKEFG0MmMcP",
  publicKey:
    "MIICCgKCAgEApq21wR1t9ZnC0R2vgmUrDGjq0G5h5THrtC4kdKfavu2t7iwV0RHsQCJU_uboan63dQysFlYLrqvUluMKBgb8lNAGRc1M7LvGKh3zQlV1gtkh71oVlRqUpuGvY6FbBSCVZThC3oRXSlKpRXSMuHY_ohmsXSkZCDReTZevLmzTG7r_bDp8mK05moNU2dfBhh7KtTssyp6etufDrpxsBXF2i-2VzhwMX2GZJY_7DouXgHWTsLIOLqH18gNMUU8RYULNqiA8vBsubVWV9ygr6wLxdZslm4yAXfMVPj6dQ7FCwo4L7gxtqK3M5Olv1C7rRi0NpTCFuUZlOnufWcrSToPKRDvAFUkZPyk9qzyiww5ckVW3ZRaPpVF36RJN-VEU2dskfw2Md2NrMV3W8LNhXryU004CSjJy6bqmer2cQ9UwbZR8y9n4Vxf45lntZMaZ3ZOYnaXuHZL-j3AcZ82bRsApcQneRhJHElA-nJicuZ4s22-B_rvOBu8YNOR5PutAGZz7J5hGr8L-wGDHktqG0ChY30RRlbvK94XitwzXljhBeDmf5SRRoUVbba8PcF91c7IHlb5q4wWxOgEXM9kb0_Oik6iXCGfZVog8DB8uYI6P2Ta5pvMY_Jt_5U-ULu3P8MPNsfHtY0n1Kpf6MP6srdz2kZupT9VL_mReSv3korv7XHMCAwEAAQ",
  type: "RSA",
};

export const TEST_RSA_KEY = Kryptos.from("b64", {
  ...RSA,
  algorithm: "RSA-OAEP-256",
  operations: ["encrypt", "decrypt"],
  use: "enc",
});
