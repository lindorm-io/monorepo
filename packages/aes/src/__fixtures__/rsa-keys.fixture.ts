import { Kryptos } from "@lindorm/kryptos";

export const RSA_KEY_SET = Kryptos.from("pem", {
  id: "c2d553c6-cccf-5afd-87eb-4cd32057bff1",
  publicKey:
    "-----BEGIN RSA PUBLIC KEY-----\n" +
    "MIICCgKCAgEAsy0BEjwn7F87mZOjqRpryYWZYCftiAm4CBFXb1e/9jhOPGYTt6IL\n" +
    "iuydOwzFyE59ekYBFryybhDhk7UuHpooOweSPej9Sr/hFBTbanROGwYXZMbrGcpF\n" +
    "9DX5svuLZu9PusYyAWd4bG/pZUaSBY9dtnNUxxUx5KyatebEGy4dIUhQ+m5Rx9f0\n" +
    "ORv1XVE5AOZbtA8AzDBTGxjm7P1zEVoXeGm/byXRTGVWdGvHq1ij44xBW1LNkdVA\n" +
    "o+rKwjRJd6gL9CBjpb++tNyfy2o0Tx5zqfXDXJW5ENGwB0IIWdNQFHqdFmFEzDN+\n" +
    "O6GihaXstIxloeBKWXkxerSVC9XBo/CvXbIUleLxisko3/mb8ZQ9aU2M65G3zW7e\n" +
    "8bSpH7lfT6CEgXr3Fssxn4EFmbdt7KDJfqIQ+hL34GsCahzJnsCzT1cCqwcg+fDF\n" +
    "sfD/J8o9XSJJlk4Smwy2zN6rghbD6f7Rclsl2S3BPdW4rXLj69b7kbu1d0gSR7AW\n" +
    "eFMmUzCMvwnxq1+ZrQ/CfkH1UYq4uYaoqTLVtUHB8Or3U5RTZ/87tm8kup1edwLC\n" +
    "QVTkiZUQVGXOeMsj53APsAtfFcSEnWDPNbl9Kp0a57Os7qTslF80Dn76fbW2v5Ro\n" +
    "4/Gay9SADliUg1/ddmOJ+eTyEhTn5CxAHAWHu/slGtQ1pmBOTugS/9UCAwEAAQ==\n" +
    "-----END RSA PUBLIC KEY-----\n",
  privateKey:
    "-----BEGIN RSA PRIVATE KEY-----\n" +
    "MIIJKAIBAAKCAgEAsy0BEjwn7F87mZOjqRpryYWZYCftiAm4CBFXb1e/9jhOPGYT\n" +
    "t6ILiuydOwzFyE59ekYBFryybhDhk7UuHpooOweSPej9Sr/hFBTbanROGwYXZMbr\n" +
    "GcpF9DX5svuLZu9PusYyAWd4bG/pZUaSBY9dtnNUxxUx5KyatebEGy4dIUhQ+m5R\n" +
    "x9f0ORv1XVE5AOZbtA8AzDBTGxjm7P1zEVoXeGm/byXRTGVWdGvHq1ij44xBW1LN\n" +
    "kdVAo+rKwjRJd6gL9CBjpb++tNyfy2o0Tx5zqfXDXJW5ENGwB0IIWdNQFHqdFmFE\n" +
    "zDN+O6GihaXstIxloeBKWXkxerSVC9XBo/CvXbIUleLxisko3/mb8ZQ9aU2M65G3\n" +
    "zW7e8bSpH7lfT6CEgXr3Fssxn4EFmbdt7KDJfqIQ+hL34GsCahzJnsCzT1cCqwcg\n" +
    "+fDFsfD/J8o9XSJJlk4Smwy2zN6rghbD6f7Rclsl2S3BPdW4rXLj69b7kbu1d0gS\n" +
    "R7AWeFMmUzCMvwnxq1+ZrQ/CfkH1UYq4uYaoqTLVtUHB8Or3U5RTZ/87tm8kup1e\n" +
    "dwLCQVTkiZUQVGXOeMsj53APsAtfFcSEnWDPNbl9Kp0a57Os7qTslF80Dn76fbW2\n" +
    "v5Ro4/Gay9SADliUg1/ddmOJ+eTyEhTn5CxAHAWHu/slGtQ1pmBOTugS/9UCAwEA\n" +
    "AQKCAgABLU4A/ykp+dQr5ZbhIP3NRJXl1k1Oo/W0A90lq5dkuCj3Z5pAjg1yW6yK\n" +
    "u0I/Vxim6XfPXOz1ISJCNKH5bHzQKnQQ+QtU5t7FIX/xepMaxVb4//+CW1xLRuF9\n" +
    "hf577h2BdhTVAod9ety3IBgMdk0gyjAn4KFDxch8YmguX4asdcumYXwjCm2dxGu/\n" +
    "sa1krp6yl5ijPrwFBQxbSSzs08EsttEdm0t9K3s+Z4jXvMzSzVSOk4QrLZHvoba0\n" +
    "QOCsqwttDpto9tH7YSmhjfCLBoGu/6DJZmhhTmdQHORuJp+zirymbXwktT/ih2TI\n" +
    "aPcdloCCaDVE0jF7g10pRz2u5IFaYuU4G1hj60KfnjVjCmZ3R6dvchnJjREgwwv4\n" +
    "/XNLbBdL/kxIFgrilh051yprVUP6eLhADmLNgw4BF4Pm1Ovrdskbyd4VxS0PQHKj\n" +
    "Q9U5bmg30B40UiwISZDrv6B+K0Qh2vMdj3mGMfJQOmgydhpERCi+SnlKIyNDMS8n\n" +
    "udpsM3GqXn4zWP8Tu1I8XtedxcEKBkULzdpqh7EBuhtBDK0FIH5ouHfD+J17G2mf\n" +
    "Ojh2ffuML0ncLT6lUG+TQQQ2cFXakVBr5JX2h0grRWyZePhsMt53G7gxo5v7nRHL\n" +
    "W6mApcPNRpYOhUqhoG0FB+12p9bMHoBRl07NHwAS5iKUVlqkkQKCAQEA2JvCPylX\n" +
    "XDN8RDVbG/XA39Rz1fannSezKC/G9jeZBoztbQvPK/gmFroQOr+FXSK699oY06Ta\n" +
    "ASkY1Fbsxd6Cgy3/02PfOSXI5zoEqPammzXZOYH31yeumC8KOcj/8Dr2GBGQ/LuM\n" +
    "xyhMW4ifsx8uRcrvSWwfbGOAzVLggkDft3koIMxt1zGjtUZJb49AqV1TRzeJAWhX\n" +
    "1zkGVfNlpkdg/XWihmH4KNs7yC8fcLD/KjuY0gbYuBJcHGS5Bs+ICa+G+rM2PE+c\n" +
    "UvpikCAwLpVDjLMzR+Lthe3KgZTeAUktQ93+UPiHMlodcTRmEx8/ub6pPq1z6tw2\n" +
    "cn9UFOFvof+Z3QKCAQEA08KQG5txxm3hR/sanEC2VOoP8brQRnWi8SEEqgKfelHq\n" +
    "hUgdgYI+P0tERh2a47nzHqKlZ0AOhq4eIYB2noKCE2p6dFWNVLclGBuIiVj9gtNi\n" +
    "fhs6n37KEqCOImlYNkAwGOA+U3ZHOavtTJ92xvVxLo0zIH8kJSJE3CRf8bObnEml\n" +
    "WAiCmPHxzTks7a23FU3Yj7OJYAh/JUONslgvlUVTDqbwUim5yxCD+klmqvL6lHOO\n" +
    "Th907/4vUEUtHStyYCq9zyYA3JGLdIBEXpzhUuAawR8qNSfAVXhKdza75lOJNb8e\n" +
    "P8F9G7BbFII7WcvdDqGVrxpzGd0BCxLZwJXrgPVqWQKCAQA1R1qoHosWarOOM3GN\n" +
    "lw0yARfc8SaZSUKQB/Bwb4x8R/2+bMEwl4EG5g1H5gPvvLMwc7zXkhztXCZ3VEo5\n" +
    "AVoo2g+q903YyKp59ZNf3Atu8ffUZVuI2IHhfPCO9hm+BUKy+N8oyxj9g7QEb76Y\n" +
    "WHsRP7FL5gR0rKTp3BDM+RgOEQcFykLaQ6N42ykE/79v6Tyqr8ZLSlcQ4T3ltlCy\n" +
    "baw1q9GmLELWCdbSjoS1O1QIF1CnmHeGDz4I3d2FB1b2yu3wjFQy0YIqBtxnhwNS\n" +
    "8fQDj7B1ZiP2ZV8CMJPKzS0NHHWJhRf8Ep5bUgApg+4wyeznNdLlWfUqhJcx66GB\n" +
    "v1UhAoIBAGDFD8/d6E9ylTmssYtoMyp40BJ6eYVJas3gdd4vVquyVYbU3s75MvL/\n" +
    "tVKJYufgaQ696vJVHo5DcXeD2jEzmLfAt/cDxZNKIoRyyKFjbm5jpSXV0aNjcHq0\n" +
    "rl9hXrNj/9bOWw0PBxt+yvFyCfh/31oToqKeF2l00Jz9bm7o0hrIuOuS8ANuxT4w\n" +
    "e6yd7wWoTmfU9tlQ6LGhvlYa4JvJU1xlw3r4oHQAtx8gfe9vHD3M1OI0k968fOik\n" +
    "xCoJWbpCrfzE9C3XwS/bEeNes8WIrhFE6afi44zrbLQv6gwmbpSGIw+ln0ORAv5L\n" +
    "SU4O0JVIGDwtzR2rljd2RNYybK3A+PkCggEBANVc5xpJvxFhSzq5S+Di9LHbsHgi\n" +
    "YgSPCbs7bkM0reykJveGRA4tCSDkxE+0nXKf6tv781REI/AWx61CHbIQMQXriHOI\n" +
    "4jDOZVK9yNQ7b6QumMZ76qMesSf8wFErjvApBp5Yh9AA4FAPKb/dKbXVIvG1MYxJ\n" +
    "vAKgwCacFXPGPjNcpFhGWXQBrrz3ac6JldSzD+o5TPfPsUKPZhV4cDuphKQApP8Y\n" +
    "B/er8Oi8/5Z6rAjY9T1dq3rSvzDJnst0lqsNP2jHMzNhAl06T517rDLmmZIUy3Cd\n" +
    "6jwDnME2QhASFRZWssA3XsgGT4lZV5NCOgLMCZiwyWAcMhhRA5t5b83waWg=\n" +
    "-----END RSA PRIVATE KEY-----\n",
  type: "RSA",
});
