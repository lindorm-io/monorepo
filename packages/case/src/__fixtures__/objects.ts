export const TEST_OBJECT = {
  camelCase: true,
  "Capital Case": true,
  CONSTANT_CASE: true,
  "dot.case": true,
  "kebab-case": true,
  "lower case": true,
  PascalCase: true,
  "path/case": true,
  "Sentence case": true,
  snake_case: true,
};

export const TEST_ARRAY_WITH_OBJECTS = [TEST_OBJECT, TEST_OBJECT];
