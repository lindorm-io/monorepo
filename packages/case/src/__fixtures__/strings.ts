export const TEST_STRINGS = [
  "camelCase",
  "Capital Case",
  "CONSTANT_CASE",
  "dot.case",
  "kebab-case",
  "lower case",
  "PascalCase",
  "path/case",
  "Sentence case",
  "snake_case",
];

export const INVALID_INPUT: Array<{ type: string; value: any }> = [
  { type: "number", value: 1234 },
  { type: "error", value: new Error("error") },
  { type: "date", value: new Date() },
  { type: "null", value: null },
  { type: "undefined", value: undefined },
];
