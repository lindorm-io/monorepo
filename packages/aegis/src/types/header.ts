import { PublicEncryptionJwk } from "@lindorm/aes";
import { KryptosEncryption, KryptosJwk } from "@lindorm/kryptos";
import { TOKEN_HEADER_ALGORITHMS, TOKEN_HEADER_TYPES } from "../constants/private";

export type TokenHeaderAlgorithm = (typeof TOKEN_HEADER_ALGORITHMS)[number];

export type TokenHeaderType = (typeof TOKEN_HEADER_TYPES)[number];

// https://www.rfc-editor.org/rfc/rfc7515.html#section-4.1
export type TokenHeaderClaims = {
  alg: TokenHeaderAlgorithm; // algorithm
  crit?: Array<Exclude<keyof TokenHeaderClaims, "crit">>;
  cty?: string; // content type
  enc?: KryptosEncryption; // encryption
  epk?: PublicEncryptionJwk; // public encryption jwk
  hkdf_salt?: string; // salt
  iv?: string; // public encryption iv
  jku?: string; // jwks uri
  jwk?: KryptosJwk; // jwk
  kid?: string; // key id
  oid?: string; // object id
  p2c?: number; // p2c
  p2s?: string; // p2s
  tag?: string; // public encryption tag
  typ: TokenHeaderType; // header type
  x5c?: Array<string>;
  x5t?: string;
  x5u?: string;
  "x5t#S256"?: string;
};

export type RawTokenHeaderClaims = {
  alg?: TokenHeaderAlgorithm;
  crit?: Array<Exclude<keyof TokenHeaderSignOptions, "critical">>;
  cty?: string;
  enc?: KryptosEncryption;
  epk?: PublicEncryptionJwk;
  hkdf_salt?: Buffer;
  iv?: Buffer;
  jku?: string;
  jwk?: KryptosJwk;
  kid?: string;
  oid?: string;
  p2c?: number;
  p2s?: Buffer;
  tag?: Buffer;
  typ?: TokenHeaderType;
  x5c?: Array<string>;
  x5t?: string;
  x5u?: string;
  "x5t#S256"?: string;
};

export type DecodedTokenHeader = TokenHeaderClaims;

export type ParsedTokenHeader = {
  algorithm: TokenHeaderAlgorithm;
  contentType: string | undefined;
  critical: Array<Exclude<keyof ParsedTokenHeader, "critical">>;
  encryption: KryptosEncryption | undefined;
  headerType: TokenHeaderType;
  hkdfSalt: string | undefined;
  jwk: KryptosJwk | undefined;
  jwksUri: string | undefined;
  keyId: string | undefined;
  objectId: string | undefined;
  pbkdfIterations: number | undefined;
  pbkdfSalt: string | undefined;
  publicEncryptionIv: string | undefined;
  publicEncryptionJwk: PublicEncryptionJwk | undefined;
  publicEncryptionTag: string | undefined;
  x5c: Array<string> | undefined;
  x5t: string | undefined;
  x5u: string | undefined;
  x5tS256: string | undefined;
};

export type TokenHeaderSignOptions = {
  algorithm?: TokenHeaderAlgorithm;
  contentType?: string;
  critical?: Array<Exclude<keyof TokenHeaderSignOptions, "critical">>;
  encryption?: KryptosEncryption;
  headerType?: TokenHeaderType;
  hkdfSalt?: Buffer;
  jwk?: KryptosJwk;
  jwksUri?: string;
  keyId?: string;
  objectId?: string;
  pbkdfIterations?: number;
  pbkdfSalt?: Buffer;
  publicEncryptionIv?: Buffer;
  publicEncryptionJwk?: PublicEncryptionJwk;
  publicEncryptionTag?: Buffer;
  x5c?: Array<string>;
  x5t?: string;
  x5u?: string;
  x5tS256?: string;
};
