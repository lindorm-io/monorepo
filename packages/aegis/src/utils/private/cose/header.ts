import { Dict } from "@lindorm/types";
import { COSE_ALGORITHM, COSE_HEADER } from "../../../constants/private";
import { AegisError } from "../../../errors";
import { RawTokenHeaderClaims } from "../../../types";
import { fromBstr, toBstr } from "./bstr";
import { decodeCoseCrit, mapCoseCrit } from "./crit";
import { findCoseByKey, findCoseByLabel } from "./find";
import { decodeCoseKey, mapCoseKey } from "./key";

export const mapCoseHeader = (claims: RawTokenHeaderClaims): Dict => {
  const result: Dict = {};

  for (const [key, value] of Object.entries(claims)) {
    const claim = findCoseByKey(COSE_HEADER, key);

    if (!claim) {
      result[key] = value;
      continue;
    }

    if (key === "alg") {
      const alg = findCoseByKey(COSE_ALGORITHM, value);
      if (!alg) {
        console.error("map", { key, value, claim, alg });
        throw new AegisError(`Unsupported COSE algorithm: ${value}`);
      }
      result[claim.label] = alg.label;
      continue;
    }

    if (key === "crit") {
      result[claim.label] = mapCoseCrit(value);
      continue;
    }

    if (key === "epk" || key === "jwk") {
      result[claim.label] = mapCoseKey(value);
      continue;
    }

    result[claim.label] = toBstr(claim, value);
    continue;
  }

  return result;
};

export const decodeCoseHeader = (cose: Dict): RawTokenHeaderClaims => {
  const result: Dict = {};

  for (const [label, value] of Object.entries(cose)) {
    const claim = findCoseByLabel(COSE_HEADER, label);

    if (!claim) {
      result[label] = value;
      continue;
    }

    if (claim.key === "alg") {
      const alg = findCoseByLabel(COSE_ALGORITHM, value);
      if (!alg) {
        console.error("decode", { label, value, claim, alg });
        throw new AegisError(`Unsupported COSE algorithm: ${value}`);
      }
      result[claim.key] = alg.key;
      continue;
    }

    if (claim.key === "crit") {
      result[claim.key] = decodeCoseCrit(value);
      continue;
    }

    if (claim.key === "epk" || claim.key === "jwk") {
      result[claim.key] = decodeCoseKey(value);
      continue;
    }

    result[claim.key] = fromBstr(claim, value);
    continue;
  }

  return result as RawTokenHeaderClaims;
};
