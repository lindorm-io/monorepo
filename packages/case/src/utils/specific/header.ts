import { KeysInput } from "../../types";
import { convertArray } from "../private";
import { headerCase } from "../private/convert-case";
import { convertObject } from "../private/convert-object";

export { headerCase };

export const headerKeys = <T extends KeysInput = KeysInput>(input: T): T =>
  convertObject(input, headerCase);

export const headerArray = (input: Array<string>): Array<string> => convertArray(input, headerCase);
