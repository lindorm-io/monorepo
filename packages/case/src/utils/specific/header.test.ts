import { TEST_ARRAY_WITH_OBJECTS, TEST_OBJECT } from "../../__fixtures__/objects";
import { INVALID_INPUT, TEST_STRINGS } from "../../__fixtures__/strings";
import { headerArray, headerCase, headerKeys } from "./header";

describe("headerCase", () => {
  test.each(TEST_STRINGS)("should convert %s", (input) => {
    expect(headerCase(input)).toMatchSnapshot();
  });

  test("should convert object keys", () => {
    expect(headerKeys(TEST_OBJECT)).toMatchSnapshot();
  });

  test("should convert array with objects", () => {
    expect(headerKeys(TEST_ARRAY_WITH_OBJECTS)).toMatchSnapshot();
  });

  test("should convert array with strings", () => {
    expect(headerArray(TEST_STRINGS)).toMatchSnapshot();
  });

  test.each(INVALID_INPUT)("should throw for invalid input $type", ({ value }) => {
    expect(() => headerCase(value)).toThrow(Error);
  });
});
