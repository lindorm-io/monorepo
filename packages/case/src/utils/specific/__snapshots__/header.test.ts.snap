// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`headerCase should convert CONSTANT_CASE 1`] = `"Constant-Case"`;

exports[`headerCase should convert Capital Case 1`] = `"Capital-Case"`;

exports[`headerCase should convert PascalCase 1`] = `"Pascal-Case"`;

exports[`headerCase should convert Sentence case 1`] = `"Sentence-Case"`;

exports[`headerCase should convert array with objects 1`] = `
[
  {
    "Camel-Case": true,
    "Capital-Case": true,
    "Constant-Case": true,
    "Dot-Case": true,
    "Kebab-Case": true,
    "Lower-Case": true,
    "Pascal-Case": true,
    "Path-Case": true,
    "Sentence-Case": true,
    "Snake-Case": true,
  },
  {
    "Camel-Case": true,
    "Capital-Case": true,
    "Constant-Case": true,
    "Dot-Case": true,
    "Kebab-Case": true,
    "Lower-Case": true,
    "Pascal-Case": true,
    "Path-Case": true,
    "Sentence-Case": true,
    "Snake-Case": true,
  },
]
`;

exports[`headerCase should convert array with strings 1`] = `
[
  "Camel-Case",
  "Capital-Case",
  "Constant-Case",
  "Dot-Case",
  "Kebab-Case",
  "Lower-Case",
  "Pascal-Case",
  "Path-Case",
  "Sentence-Case",
  "Snake-Case",
]
`;

exports[`headerCase should convert camelCase 1`] = `"Camel-Case"`;

exports[`headerCase should convert dot.case 1`] = `"Dot-Case"`;

exports[`headerCase should convert kebab-case 1`] = `"Kebab-Case"`;

exports[`headerCase should convert lower case 1`] = `"Lower-Case"`;

exports[`headerCase should convert object keys 1`] = `
{
  "Camel-Case": true,
  "Capital-Case": true,
  "Constant-Case": true,
  "Dot-Case": true,
  "Kebab-Case": true,
  "Lower-Case": true,
  "Pascal-Case": true,
  "Path-Case": true,
  "Sentence-Case": true,
  "Snake-Case": true,
}
`;

exports[`headerCase should convert path/case 1`] = `"Path-Case"`;

exports[`headerCase should convert snake_case 1`] = `"Snake-Case"`;
